packages/panzer/adapters-ioss/src/Panzer_IOSSConnManager.cpp

```diff
@@ -40,21 +40,970 @@
+                                  GlobalOrdinal & faceOffset, GlobalOrdinal & cellOffset) const {
+
+  GlobalOrdinal nodeId;
+  GlobalOrdinal blockmaxNodeId;
+  GlobalOrdinal localmaxNodeId = -1;
+  GlobalOrdinal maxNodeId = -1;
+
+  GlobalOrdinal maxEdgeId = -1;
+  GlobalOrdinal maxFaceId = -1;
+
+  GlobalOrdinal blockmaxElementId;
+  GlobalOrdinal localmaxElementId = std::numeric_limits<GlobalOrdinal>::min();
+  GlobalOrdinal maxElementId;
+
+  const CellTopologyData * baseTopologyData = fp.getCellTopology().getBaseCellTopologyData();
+
+  int nodesPerElement = fp.getSubcellCount(0);
+  int cornerNodesPerElement = int(baseTopologyData->subcell_count[0]);
+  bool extended = nodesPerElement > cornerNodesPerElement;
+
+  // get the local maximum node, edge, and face numbers
+  std::vector<std::string> elementBlockIds;
+  getElementBlockIds(elementBlockIds);
+
+  int elementsInBlock;
+  Teuchos::RCP<std::vector<GlobalOrdinal>> blockConnectivity;
+  for (std::string elementBlockId : elementBlockIds) {
+	std::vector<LocalOrdinal> elementBlock = *(elementBlocks_.find(elementBlockId)->second);
+	blockmaxElementId = -1;
+	for (auto element : elementBlock) {
+	  if (element > blockmaxElementId) {
+	    blockmaxElementId = elmtLidToGid_[element];
+	  }
+	}
+	blockConnectivity = iossConnectivity_.find(elementBlockId)->second;
+	int blockConnectivitySize = blockConnectivity->size();
+	if (blockConnectivitySize > 0) {
+	  const Ioss::ElementTopology * iossElementBlockTopology = iossElementBlockTopologies_.find(elementBlockId)->second;
+	  int iossNodesPerElement = iossElementBlockTopology->number_nodes();
+	  if (!extended && (cornerNodesPerElement < iossNodesPerElement)) {
+	    // If the FieldPattern is not extended, but the Ioss::ElementTopology is, then
+		// we must ignore the non-corner nodes.
+		elementsInBlock = blockConnectivitySize / iossNodesPerElement;
+		TEUCHOS_TEST_FOR_EXCEPTION(blockConnectivitySize % iossNodesPerElement != 0, std::logic_error,
+		   "Error, Size of connectivity for this block not a multiple of the number of elements in the block.");
+		blockmaxNodeId = -1;
+		for (int element = 0; element < elementsInBlock; ++element) {
+		  for (int node = 0; node < cornerNodesPerElement; ++node) {
+		    nodeId = *(blockConnectivity->begin() + nodesPerElement*element + iossElementBlockTopology->element_connectivity()[node]);
+		    if (nodeId > blockmaxNodeId) {
+		      blockmaxNodeId = nodeId;
+		    }
+		  }
+		}
+	  }
+	  else {
+	    // We can use the entire blockConnectivity vector if the FieldPattern is extended
+		// or if neither the FieldPattern nor the Ioss::ElementTopology is extended.
+		blockmaxNodeId = *(std::max_element(blockConnectivity->begin(), blockConnectivity->end()));
+	  }
+          if (blockmaxNodeId > localmaxNodeId)
+            localmaxNodeId = blockmaxNodeId;
+	}
+	if (blockmaxElementId > localmaxElementId)
+          localmaxElementId = blockmaxElementId;
+  }
+
+#ifdef HAVE_MPI
```

Comment:
See above.  Also, you should merge the two all-reduces into a single all-reduce of an array of length two.
```c++
std::array<GlobalOrdinal, 2> lclMaxes {localmaxNodeId, localmaxElementId};
std::array<GlobalOrdinal, 2> gblMaxes;
Teuchos::reduceAll<int, GlobalOrdinal> (comm, Teuchos::REDUCE_MAX, 2, lclMaxes.data (), gblMaxes.data ());
```
Panzer is really worried about strong scaling these days.  If that's a thing, then you need to avoid latency-bound MPI communication wherever you can.